### PR TITLE
THRIFT-4633: .net framework 4.5 Async Client example

### PR DESCRIPTION
**Pull request by kaanid**
Some helpful tips for a successful Apache Thrift PR:

* Did you test your changes locally or using CI in your fork?
* Is the Apache Jira THRIFT ticket identifier in the PR title?
* Is the Apache Jira THRIFT ticket identifier in the commit message?
* Did you squash your changes to a single commit?
* Are these changes backwards compatible? (please say so in PR description)
* Do you need to update the language-specific README?

Example ideal pull request title:

        THRIFT-9999: an example pull request title

Example ideal commit message:

        THRIFT-9999: [summary of fix, one line if possible]
        Client: [language(s) affected, comma separated, use lib/ directory names please]

For more information about committing, see CONTRIBUTING.md

### CODE REVIEW

File: tutorial/csharp.async/tutorial.thrift [hunk 1]
```diff
@@ -0,0 +1,156 @@
+/*
```
Comment (jeking3) [inline tutorial/csharp.async/tutorial.thrift]: This looks like a duplicate of the same file located elsewhere?

File: tutorial/csharp.async/Tutorial/obj/Debug/Tutorial.csproj.FileListAbsolute.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+F:\thrift\2\thrift\tutorial\csharp.async\Tutorial\obj\Debug\Tutorial.csprojAssemblyReference.cache
```
Comment (jeking3) [inline tutorial/csharp.async/Tutorial/obj/Debug/Tutorial.csproj.FileListAbsolute.txt]: Tutorial/obj should not be part of the PR.

File: tutorial/csharp.async/Tutorial/Work.cs [hunk 3]
```diff
@@ -0,0 +1,249 @@
+/**
```
Comment (jeking3) [inline tutorial/csharp.async/Tutorial/Work.cs]: Generated thrift files should not be part of the PR.

Comment (kaanid) [reply tutorial/csharp.async/Tutorial/Work.cs]: delete files.

File: tutorial/csharp.async/Tutorial/InvalidOperation.cs [hunk 4]
```diff
@@ -0,0 +1,165 @@
+/**
+ * Autogenerated by Thrift Compiler (0.11.0)
```
Comment (jeking3) [inline tutorial/csharp.async/Tutorial/InvalidOperation.cs]: Generated code should not be checked in.

File: tutorial/csharp.async/Tutorial/Calculator.cs [hunk 5]
```diff
@@ -0,0 +1,1346 @@
+/**
+ * Autogenerated by Thrift Compiler (0.11.0)
```
Comment (jeking3) [inline tutorial/csharp.async/Tutorial/Calculator.cs]: Generated code should not be checked in.

File: tutorial/csharp.async/Tutorial/Operation.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Autogenerated by Thrift Compiler (0.11.0)
```
Comment (jeking3) [inline tutorial/csharp.async/Tutorial/Operation.cs]: Generated code should not be checked in.

File: tutorial/csharp.async/Tutorial/SharedService.cs [hunk 7]
```diff
@@ -0,0 +1,499 @@
+/**
+ * Autogenerated by Thrift Compiler (0.11.0)
```
Comment (jeking3) [inline tutorial/csharp.async/Tutorial/SharedService.cs]: Generated code should not be checked in.

File: tutorial/csharp.async/Tutorial/SharedStruct.cs [hunk 8]
```diff
@@ -0,0 +1,162 @@
+/**
+ * Autogenerated by Thrift Compiler (0.11.0)
```
Comment (jeking3) [inline tutorial/csharp.async/Tutorial/SharedStruct.cs]: Generated code should not be checked in.

File: tutorial/csharp.async/Tutorial/Work.cs (hunk 3 above)
Comment (jeking3) [reply tutorial/csharp.async/Tutorial/Work.cs]: This looks like it's still here.

File: tutorial/csharp.async/Tutorial/tutorial.Constants.cs [hunk 9]
```diff
@@ -0,0 +1,30 @@
+/**
+ * Autogenerated by Thrift Compiler (0.11.0)
```
Comment (jeking3) [inline tutorial/csharp.async/Tutorial/tutorial.Constants.cs]: Generated code should not be checked in.

**Comment by kaanid** [comment]
Thank you for reviewing again.

**Comment by kaanid** [comment]
Thank you for reviewing again.

**Comment by Jens-G** [comment]
Looking into it ...

File: tutorial/csharp.async/CSharpAsyncTutorial.Server/Properties/AssemblyInfo.cs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("CSharpAsyncTutorial.Server")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Microsoft")]
+[assembly: AssemblyProduct("CSharpAsyncTutorial.Server")]
+[assembly: AssemblyCopyright("Copyright © Microsoft 2018")]
```
Comment (Jens-G) [inline tutorial/csharp.async/CSharpAsyncTutorial.Server/Properties/AssemblyInfo.cs]: (c) ASF ... etc

File: tutorial/csharp.async/README.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
+# Tutorial Frameword 4.5+ Clent Async
+
+## Setp
+1. thrift -gen csharp:async shared.thrift
+2. thrift -r -gen csharp:async tutorial.thrift
```
Comment (Jens-G) [inline tutorial/csharp.async/README.md]: thrift -r generates the file AND all included files. Is step 1 really necessary then?

File: tutorial/csharp.async/README.md [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+3. Put it in Tutorial folder
+4. Generated && Run
```
Comment (Jens-G) [inline tutorial/csharp.async/README.md]: generate, not generated

File: tutorial/csharp.async/Tutorial/Properties/AssemblyInfo.cs [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("Tutorial")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Microsoft")]
+[assembly: AssemblyProduct("Tutorial")]
+[assembly: AssemblyCopyright("Copyright © Microsoft 2018")]
```
Comment (Jens-G) [inline tutorial/csharp.async/Tutorial/Properties/AssemblyInfo.cs]: company, product,copyright ...

File: tutorial/csharp.async/README.md (hunk 12 above)
Comment (Jens-G) [reply tutorial/csharp.async/README.md]: I have a general problem wth the workflow. IMHO even if we copy the files manually, we should keep the folder structure. Hence, I'd suggest to have the user copy the entire gen-csharp folder as is, not single files.

**Comment by Jens-G** [comment]
I get lots of errors in Calculator.cs?

**Comment by kaanid** [comment]
thrift -r --gen csharp:async tutorial.thrift 
There is a problem that does not generate shared.thrift async method.
so,Secondary generation
==============================================
thrift --gen csharp:async tutorial.thrift 
There is still a problem, the asynchronous method inherits the error and needs to be manually modified.

See README.md for details.

**Comment by jeking3** [comment]
Based on the comments it looks like more work is needed and this needs to be rebased.  Please complete this.

File: tutorial/csharp.async/README.md [hunk 14]
```diff
@@ -0,0 +1,12 @@
+# Tutorial Frameword 4.5+ Clent Async
```
Comment (jeking3) [inline tutorial/csharp.async/README.md]: spelling: Framework

**Comment by Jens-G** [comment]
Not sure what we should do with this … still full of issues and C# is planned to be replaced by netstd anyways.  Are we sure we get it up and running before I'm through with netstd? Otherwise I would suggest to just cancel this PR.

**Comment by kaanid** [comment]
thx,
I learned a lot from this RP.
I suggest you use dotnet.
